### DISCUSSION TITLE
How to stop channel processing on encountering an error in input message ?

### DISCUSSION THREAD

**Discussion by sanjay86alld** [nextgenhealthcare/connect#4931] [Q&A]
Hello All ,

In my channel destination transformer I have written script to convert input messages to output.
I am writing the code in try catch block. On encountering an error in input message I am trying to stop the channel from further processing via this statement in catch block -

ChannelUtil.stopChannel(channelId);

My question is - This is not working as expected . I see the channel status changing from "Started" to "Stopping" to "Stopped" status but only after a while. By that time it has processed few more messages after the message which was having error.
I want to stop the processing as soon as the we get a message which has error .

I tried this as well - ChannelUtil.haltChannel(channelId); ( using ) - This is also giving same behavior .

How can I stop the channel from further processing any more messages ? 
Any other approach which I should take here .

**Comment by pacmano1** [comment]
Are you batching in any way?

**Reply by pacmano1** [reply]
Also I would probably not implement a halt at all.   What is the use case here?   Curious.  I might expect bypassing the record on error or some other mechanism such that hard stops are never required - e.g. send to another channel that returns a condition such that your initial channels exits its iteration, but does not stop or halt itself.

**Reply by sanjay86alld** [reply]
Yes doing a batch processing.

Use case is - We need to convert an EDI 834 file to flat file layout. We want to stop processing if we get an invalid edi transaction ( like having invalid date or count mismatch ) and raise an incident ticket.
Next thing should be - we need to correct the input data and resume the processing.
 
Issue we are getting here is - on encountering the error the channel is not stopping instantly but takes some time and processes few more messages before it stops .

**Reply by tonygermano** [reply]
Are you able to share your channel? I'm not entirely clear on how you're batch processing. In general, it's not possible to stop in the middle of a batch and then resume where you left off.

**Comment by odoodo** [comment]
Try the following:
1. choose "split Batch by" "JavaScript" under Set Data Types/Inbound Properties
2. create javascript that provides 1 line every call if not stop processing attribute has been set in global channel map: 
```
var nextMessage = null;
// if processing has not been explicitly stopped and if there is something more to read
if(!globalChannelMap.containsKey('stopProcessing') && (nextMessage = reader.readLine())){
   // also add a linebreak if needed
    nextMessage += '\n';
}
// and provide the result to the channel, if any
return nextMessage;
```

3. within the catch block in your channel, simply set the attribute in the global channel map that tells the reader to stop its work:
```
try{
...
}catch(transactionError){ 
    // tell the reader in the connector to not go on w/ its work
    globalChannelMap.set('stopProcessing', true);
    // don't even go on to process this message
    destinationSet.removeAll();
    return;
}
```
_I am not 100% sure if the batch splitter & channel are working synchronously. So it might not work like intended if they are not. In this (quite unlikely) case, you would have to check the existence of 'stopProcessing' in the source filter of your channel._
